# score-assemblies

A snakemake-wrapper for evaluating *de novo* bacterial genome assemblies, e.g. from Oxford Nanopore (ONT) or Illumina sequencing.

The workflow includes the following programs:
* [pomoxis](https://github.com/nanoporetech/pomoxis) assess_assembly and assess_homopolymers
* dnadiff from the [mummer](https://mummer4.github.io/index.html) package
* [NucDiff](https://github.com/uio-cels/NucDiff/)
* [QUAST](http://quast.sourceforge.net/quast)
* [BUSCO](https://busco.ezlab.org/)
* [ideel](https://github.com/mw55309/ideel/), which uses [prodigal](https://github.com/hyattpd/Prodigal) and [diamond](https://github.com/bbuchfink/diamond)
* [bakta](https://github.com/oschwengers/bakta)

## Installation
Clone repository, for example:
```
git clone https://github.com/pmenzel/score-assemblies.git /opt/software/score-assemblies
```
Create a new conda environment containing all necessary programs:
```
conda env create -n score-assemblies --file /opt/software/score-assemblies/env/environment.yaml
```
and activate the environment:
```
conda activate score-assemblies
```

## Usage
First, prepare a data folder, which must contain subfolders `assemblies/` containing the
assemblies and, optionally, `references/` and `references-protein/`, which containin reference genomes and their proteins with which the assemblies and predicted proteins will be compared.  
For example:
```
.
├── assemblies
│   ├── flye-i4.fa
│   ├── flye-i4+raconx4.fa
│   ├── flye-i4+raconx4+medaka.fa
│   ├── raven-p4.fa
│   └── raven-p4+medaka.fa
│
├── references
│   └── bac_sub_ref.fa
│
└── references-protein
    └── bac_sub_ref.faa
```
NB: The fasta files need to have the `.fa` extension.

Run workflow in that folder, e.g. with 20 threads:
```
snakemake -s /opt/software/score-assemblies/Snakefile --cores 20 --use-conda
```

If no references are supplied, then only ideel and BUSCO are done.

### Modules
score-assemblies will run these programs on each assembly:

#### assess_assembly and assess_homopolymers

Each assembly will be compared against each reference genome using the
`assess_assembly` and `assess_homopolymers` scripts from
[pomoxis](https://github.com/nanoporetech/pomoxis).  Additionally to the tables
and plots generated by these programs, summary plots for each reference genome will be plotted
in `pomoxis/<reference>_assess_assembly_all_meanQ.pdf` and
`pomoxis/<reference>_assess_homopolymers_all_correct_len.pdf`.

##### Example plot assess_assembly
![Example assess_assembly](example/example_assess_assembly.png?raw=true)

##### Example plot assess_homopolymers
![Example assess_homopolymers](example/example_assess_homopolymers.png?raw=true)

#### BUSCO

Set the lineage via the snakemake call:
```
snakemake -s /opt/software/score-assemblies/Snakefile --cores 20 --config busco_lineage=bacillales
```
If not set, the default lineage `bacteria` will be used.
Available datasets can be listed with `busco --list-datasets`

The number of complete, fragmented and missing BUSCOs per assembly is plotted in `busco/busco_stats.pdf`.

##### Example plot of BUSCO results
![Example BUSCO](example/example_busco.png?raw=true)

#### dnadiff
Each assembly is compared with each reference and the output files will be
located in `dnadiff/<reference>/<assembly>-dnadiff.report`.  The values for
`AvgIdentity` (from 1-to-1 alignments) and `TotalIndels` are extracted from these files and are plotted
for each reference in `dnadiff/<reference>_dnadiff_stats.pdf`.

##### Example plot of dnadiff results
![Example dnadiff](example/example_dnadiff.png?raw=true)


#### NucDiff
Each assembly is compared with each reference and the output files will be
located in the folder `nucdiff/<reference>/<assembly>-nucdiff/`.  The values for
`Insertions`, `Deletions`, and `Substitutions` are extracted from the file `results/nucdiff_stat.out` and are plotted
for each reference in `nucdiff/<reference>_nucdiff_stats.pdf`.

#### QUAST

One QUAST report is generated for each reference genome, containing the results for all assemblies.
The report files are located in `quast/<reference>/report.html`.

#### bakta

bakta is only run when specified as extra config argument in the snakemake call:
```
snakemake -s /opt/software/score-assemblies/Snakefile --cores 20 --use-conda --config busco_lineage=bacillales bakta=1
```
NB: It takes some time to download the bakta database and run bakta on all assemblies.

#### ideel

Open reading frames are predicted from each assembly via Prodigal and are
search in the Uniprot sprot database with diamond, retaining the best alignment
for each ORF. For each assembly, the distribution of the ratios between length
of the ORF and the matching database sequence are plotted to `ideel/ideel_stats.pdf`.

Additionally, diamond alignments are done between the predicted ORFs and the supplied reference proteins.

##### Example plot of ideel result
![Example ideel](example/example_ideel.png?raw=true)

