---
title: score-assemblies
output:
  html_document:
    theme: united
    toc: yes
    toc_depth: 4
---

```{css, echo=FALSE}
    body .main-container {
      max-width: 1200px !important;
      width: 1200px !important;
    }
    
    .figure {
			margin-top: 50px;
			margin-bottom: 50px;
    }
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
library(tidyverse)
library(DT)
library(knitr)
#library(plotly)

options(dplyr.summarise.inform = FALSE)
opts_knit$set(root.dir = "~/temp/score-assemblies")

```

## BUSCO

[BUSCO](https://busco.ezlab.org/) measures the percentage of complete, fragmented, or missing core genes that are usually present in bacterial clades.

```{r busco}
filename_busco_stats_tsv <- "busco/all_stats.tsv"

df.busco <- read_tsv(filename_busco_stats_tsv, col_names = c("assembly", "% complete", "% fragmented", "% missing", "n"), col_types = "fdddd") %>%
  select(-n)

df.busco %>%
  DT::datatable(escape = FALSE, rownames = FALSE, class="compact hover",
                options = list(pageLength = 10, searchHighlight = TRUE, paginate = TRUE, lengthChange = FALSE, fixedHeader = TRUE, order = list(list(1, 'desc'), list(0, 'asc')))) %>%
  formatStyle(names(df.busco), 'font-size' = '90%') %>%
  formatStyle(names(df.busco), 'vertical-align' = 'top')
```


## pomoxis

The [pomoxis](https://github.com/nanoporetech/pomoxis) tool _assess_assembly_ compares the assembly to a reference genome resulting in a mean Q score and error rate over all sequences in the reference genome.

```{r pomoxis_assess_assembly}
filename_assess_assembly_all_scores_tsv <- "pomoxis/assess_assembly_all_scores.tsv"
df.assess_assembly <- read_tsv(filename_assess_assembly_all_scores_tsv, col_names = c("assembly", "reference", "Qscore", "percErr"), col_types = "ffdc") %>%
  mutate(percErr = parse_number(percErr))
df.assess_assembly %>%
  DT::datatable(escape = FALSE, rownames = FALSE, class="compact hover",
               options = list(pageLength = 10, searchHighlight = TRUE, paginate = TRUE, lengthChange = FALSE, fixedHeader = TRUE, order = list(list(2, 'desc'), list(0, 'asc')))) %>%
  formatStyle(names(df.assess_assembly), 'font-size' = '90%') %>%
  formatStyle(names(df.assess_assembly), 'vertical-align' = 'top')

```

The [pomoxis](https://github.com/nanoporetech/pomoxis) tool _assess_homopolymers_ compares the assembly to a reference genome and calculates the error rate for each homopolymer length.  
The table shows the rates of correct homopolymers of length 3 to 8 for the four bases A, C, G, and T as well as the mean of all values in the last column.
```{r pomoxis_assses_homopolymers}
filename_assess_homopolymers_all_correct_len_tsv <- "pomoxis/assess_homopolymers_all_correct_len.tsv"

df.assess_homopolymers <- read_tsv(filename_assess_homopolymers_all_correct_len_tsv,
  col_names = c("assembly", "reference", "rlen", "A", "C", "G", "T", "A_n", "C_n", "G_n", "T_n", "AT_n", "AT", "GC_n", "GC"),
  col_types = "fffdddddddddddd"
) %>%
  select(-ends_with("_n")) %>%
  gather(key = "base", value = "frac_correct", -assembly, -reference, -rlen)

df.single_nuc <- df.assess_homopolymers %>%
  filter(str_length(base) == 1) %>%
  mutate(rlen = paste0("len", rlen)) %>%
  group_by(assembly, reference, rlen) %>%
  summarise(mean_frac_correct  = mean(frac_correct, na.rm = TRUE)) %>%
  ungroup() %>%
  pivot_wider(id_cols = c(assembly, reference), names_from = rlen, values_from = mean_frac_correct) %>%
  rowwise() %>%
  mutate(`mean` = mean(c(len3, len4, len5, len6, len7, len8, na.rm = TRUE)))

df.single_nuc %>%
   DT::datatable(escape = FALSE, rownames = FALSE, class="compact hover",
               options = list(pageLength = 10, searchHighlight = TRUE, paginate = TRUE, lengthChange = FALSE, fixedHeader = TRUE, order = list(list(8, 'desc'), list(0, 'asc')))) %>%
  formatRound(c(3:9), digits = 3) %>%
  formatStyle(names(df.single_nuc), 'font-size' = '90%') %>%
  formatStyle(names(df.single_nuc), 'vertical-align' = 'top')

```

## IDEEL

### Comparison with Uniprot Sprot database

ORFs are predicted in each assembly using Prodigal, which are then searched against die Uniprot SPROT database using diamond.
Measured is the fraction of predicted ORFS, which were â‰¥95% the length of their best-matching protein sequence in the database (higher fractions being better).

```{r}
dir_ideel_diamond <- "ideel/diamond/"
filelist <- list.files(path = dir_ideel_diamond, pattern = ".*\\.tsv", recursive = FALSE, full.names = FALSE)

df_list <- vector("list", length(filelist))
for (i in seq_along(filelist)) {
  filename <- paste0(dir_ideel_diamond, filelist[[i]])
  df <- read_tsv(filename,
    col_names = c("qlen", "slen"),
    col_types = "ii"
  ) %>%
    mutate(assembly = filelist[[i]]) %>%
    mutate(assembly = str_remove(assembly, "\\.tsv")) %>%
    mutate(SCov = qlen / slen)

  df_list[[i]] <- df
}
df.all <- dplyr::bind_rows(df_list) %>% mutate(assembly = factor(assembly))

df.ideel_uniprot <- df.all %>%
  group_by(assembly) %>%
  summarise(
    n_ORFS = n(),
    n_full_length_ORFs = sum(qlen == slen),
    n_95perc_ORFs = sum(qlen >= 0.95 * slen)
  ) %>% 
  ungroup()

df.ideel_uniprot %>% 
 DT::datatable(escape = FALSE, rownames = FALSE, class="compact hover",
               options = list(pageLength = 10, searchHighlight = TRUE, paginate = TRUE, lengthChange = FALSE, fixedHeader = TRUE, order = list(list(3, 'desc'), list(0, 'asc')))) %>%
  formatStyle(names(df.ideel_uniprot), 'font-size' = '90%') %>%
  formatStyle(names(df.ideel_uniprot), 'vertical-align' = 'top')
```

```{r ideel_uniprot, warning=FALSE, fig.width = 12, fig.height = 10, out.width = '1200px'}
p <- ggplot(df.all, aes(x = SCov)) +
  geom_histogram(aes(color=assembly, fill=assembly), alpha = 0.8, position = "identity", binwidth = 0.01, size=0.2) +
  #geom_density(aes(fill = assembly), alpha = 0.4, color = "grey40") +
  facet_wrap(~assembly, scales = "free_x") +
  theme_bw() +
  ggtitle(paste("IDEEL Uniprot")) +
  theme(legend.position = "none") +
  ylab("") + xlab("qlen / slen") +
  theme(strip.text = element_text(size = 6)) +
  xlim(0, 1.5)
p
```

### Comparison with reference genomes

ORFs are predicted in each assembly using Prodigal, which are then searched against the provided reference protein sequences using diamond.
Measured is the number and fraction of predicted ORFS, which have the same length as their best-matching reference protein sequence.

```{r}

# TODO load number of refefence proteins from faa files and add to table

df.ideel_ref <- data.frame(assembly = character(), reference = character(), n_ORFs = integer(), n_full_length_ORFs = integer(), n_min95perc_ORFs = integer())

dir_ideel_diamond_ref <- "ideel/diamond-ref/"
ref_list <- list.dirs(path = dir_ideel_diamond_ref, recursive = FALSE, full.names = FALSE)
for (r in seq_along(ref_list)) {

	filelist <- list.files(path = paste0(dir_ideel_diamond_ref,"/", ref_list[[r]]), pattern = ".*\\.tsv", recursive = FALSE, full.names = FALSE)
	for (i in seq_along(filelist)) {
		filename <- paste0(dir_ideel_diamond_ref, "/", ref_list[[r]], "/", filelist[[i]])
		df <- read_tsv(filename, col_names = c("qlen", "slen"),	col_types = "ii")
		
		n_total <- df %>% nrow()
	  n_same_len <- df %>% filter(slen == qlen) %>% nrow()
	  n_min95perc_len <- df %>% filter(qlen >= 0.95 * slen) %>% nrow()
	  
  	assembly <- filelist[[i]] %>% str_remove("\\.tsv") %>% str_remove(paste0("_",ref_list[[r]]))
	
  	df.ideel_ref <- df.ideel_ref %>% add_row(reference = ref_list[[r]], assembly = assembly, n_ORFs = n_total, n_full_length_ORFs = n_same_len, n_min95perc_ORFs = n_min95perc_len)
	}
}

df.ideel_ref %>% 
 DT::datatable(escape = FALSE, rownames = FALSE, class="compact hover",
               options = list(pageLength = 10, searchHighlight = TRUE, paginate = TRUE, lengthChange = FALSE, fixedHeader = TRUE, order = list(list(3, 'desc'), list(0, 'asc')))) %>%
  formatStyle(names(df.ideel_ref), 'font-size' = '90%') %>%
  formatStyle(names(df.ideel_ref), 'vertical-align' = 'top')

```
