---
title: score-assemblies
output:
  html_document:
    theme: united
    toc: yes
    toc_depth: 4
    mathjax: null
---

```{css, echo=FALSE}
    body .main-container {
      max-width: 1200px !important;
      width: 1200px !important;
    }

	.figure {
			margin-top: 50px;
			margin-bottom: 50px;
    }
    /* give background color to the first two rows with th, https://stackoverflow.com/a/35252091/4598742 */
    #dt_summary table thead tr:nth-child(-n+2) th {
      background-color: #dfdfdf;
      --padding: 3px;
      padding: var(--padding);
      border: 0px solid white;
      box-shadow: inset 0 0 0 var(--padding) white;
      white-space: wrap;
    }

    #dt_summary table thead tr:nth-child(3) th {
      font-size: 90%;
      white-space: nowrap;
    }

    #dt_summary td {
      white-space: nowrap;
    }

```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
library(tidyverse)
library(DT)
library(htmltools)
library(knitr)
library(Biostrings)
#library(plotly)

options(dplyr.summarise.inform = FALSE)

green <- "100, 255, 100"
red <- "255, 100, 100"

out_dir <- args[2]

path_references <- paste0(getwd(),'/references')
path_references_protein <- paste0(getwd(),'/references-protein')
path_bakta <- paste0(out_dir,'/bakta/') # with trailing /
do_reference_genomes <- dir.exists(path_references)
do_reference_proteomes <-  dir.exists(path_references_protein)
do_bakta <-  dir.exists(path_bakta)

#for debugging:
#opts_knit$set(root.dir = "~/temp/score-assemblies")
#path_references <- paste0("~/temp/score-assemblies", '/references')
#path_references_protein <- paste0("~/temp/score-assemblies", '/references-protein')
#do_reference_genomes <- TRUE
#do_reference_proteomes <- TRUE

list_ref <- list()
list_ref_protein <- list()

if(do_reference_genomes) {
  list_ref <- list.files(path = path_references, pattern = ".*\\.fa$", recursive = FALSE, full.names = FALSE)
  list_ref <- sort(str_remove(list_ref, "\\.fa$"))
}
if(do_reference_proteomes) {
  list_ref_protein <- list.files(path = path_references_protein, pattern = ".*\\.faa$", recursive = FALSE, full.names = FALSE)
  list_ref_protein <- sort(str_remove(list_ref_protein, "\\.faa$"))
}

color_dt <- function(dt, col_range, base_colors, df) {
  alpha_green <- seq(0, 0.7, length.out = 10)
  alpha_red <- seq(0.3, 1.0, length.out = 10)
  for (i in col_range) {
    values <- na.omit(df[[i]])
    n <- length(unique(values))
    c <- base_colors[[i - (ncol(df)-length(base_colors)) ]]
    if(n == 1) { # only one value in whole column
      list_breaks <- c(values[[1]])
      list_colors <- rep(paste0("rgba(", c,",1.0)"), 2)
    } else {
      if(c == green) {
        second_highest_value <- sort(unique(values))[n-1]
        list_breaks <- seq(min(values), second_highest_value, length.out = 10)
        list_colors <- paste0("rgba(", c,", ", c(alpha_green, 1.0), ")")
      } else {
        second_lowest_value <- sort(unique(values))[2]
        list_breaks <- seq(second_lowest_value, max(values), length.out = 10) - 0.0001 # to shift the breaks to the left of the values
        list_colors <- paste0("rgba(", c,", ", c(0.0, alpha_red), ")")
      }
    }
    dt <- DT::formatStyle(dt, i, backgroundColor = DT::styleInterval(list_breaks, list_colors))
  }
  dt
}

```

```{r quasts, eval=do_reference_genomes, results='asis'}
cat("## Quast\n")
for(r in list_ref) {
  x <- paste0("* [Quast for ", r,"](", out_dir, "/quast/", r, "/report.html)\n")
  cat(x)
}
```

## BUSCO

[BUSCO](https://busco.ezlab.org/) measures the percentage of complete, fragmented, or missing core genes that are usually present in bacterial clades.

```{r busco}
filename_busco_stats_tsv <- paste0(out_dir, "/busco/all_stats.tsv")

df.busco <- read_tsv(filename_busco_stats_tsv, col_names = c("assembly", "% complete", "% fragmented", "% missing", "n"), col_types = "fdddd") %>%
  select(-n)

dt <- df.busco %>%
  DT::datatable(escape = FALSE, rownames = FALSE, class="compact hover",
                options = list(pageLength = 10, searchHighlight = TRUE, paginate = TRUE, lengthChange = FALSE, fixedHeader = TRUE, order = list(list(1, 'desc'), list(0, 'asc')))) %>%
  formatStyle(names(df.busco), 'font-size' = '90%') %>%
  formatRound(2:4, digits = 1)

color_dt(dt, 2:4, c(green, red, red), df.busco)
```

```{r pomoxis_assess_assembly, eval=do_reference_genomes}
asis_output("## pomoxis\n")

asis_output("The [pomoxis](https://github.com/nanoporetech/pomoxis) tool _assess_assembly_ compares the assembly to a reference genome resulting in a mean Q score and error rate over all sequences in the reference genome.\\n")

filename_assess_assembly_all_scores_tsv <- paste0(out_dir, "/pomoxis/assess_assembly_all_scores.tsv")
df.assess_assembly <- read_tsv(filename_assess_assembly_all_scores_tsv, col_names = c("assembly", "reference", "Qscore", "percErr"), col_types = "ffdc") %>%
  mutate(percErr = parse_number(percErr))

dt <- df.assess_assembly %>%
  DT::datatable(escape = FALSE, rownames = FALSE, class="compact hover",
               options = list(pageLength = 10, searchHighlight = TRUE, paginate = TRUE, lengthChange = FALSE, fixedHeader = TRUE, order = list(list(2, 'desc'), list(0, 'asc')))) %>%
  formatStyle(names(df.assess_assembly), 'font-size' = '90%') %>%
  formatRound(3:4, digits = 3)

color_dt(dt, 3:4, c(green, red), df.assess_assembly)
```

```{r pomoxis_assses_homopolymers, eval=do_reference_genomes}
asis_output("The [pomoxis](https://github.com/nanoporetech/pomoxis) tool _assess_homopolymers_ compares the assembly to a reference genome and calculates the error rate for each homopolymer length.  \n")
asis_output("The table shows the average rates of correct homopolymers of length 3 to 8 (longer homopolymers are ignored here) over the four bases A, C, G, and T as well as the mean of all those values in the last column.  \nNB: This mean is a bit stupid. For example, if there only two homopolymers of length 8 and one is wrong, the rate would be 0.5 for length 8 and this 0.5 will go into the mean with equal weight as the other lengths, overly drawing the mean down.  \n")
asis_output("Assembly names link to pomoxis plot for error rates of each base.\n")

filename_assess_homopolymers_all_correct_len_tsv <- paste0(out_dir, "/pomoxis/assess_homopolymers_all_correct_len.tsv")

df.assess_homopolymers <- read_tsv(filename_assess_homopolymers_all_correct_len_tsv,
  col_names = c("assembly", "reference", "rlen", "A", "C", "G", "T", "A_n", "C_n", "G_n", "T_n", "AT_n", "AT", "GC_n", "GC"),
  col_types = "ffidddddddddddd"
) %>%
  select(-ends_with("_n")) %>%
  gather(key = "base", value = "frac_correct", -assembly, -reference, -rlen)

df.single_nuc <- df.assess_homopolymers %>%
  filter(rlen <= 8) %>%
  filter(str_length(base) == 1) %>%
  #mutate(rlen = paste0("len", rlen)) %>%
  group_by(assembly, reference, rlen) %>%
  summarise(mean_frac_correct  = mean(frac_correct, na.rm = TRUE)) %>%
  ungroup() %>%
  pivot_wider(id_cols = c(assembly, reference), names_from = rlen, values_from = mean_frac_correct) %>%
  #rowwise() %>%
  #mutate(`mean` = mean(c(len3, len4, len5, len6, len7, len8), na.rm = TRUE)) %>%
  mutate(mean = rowMeans(across(where(is.numeric)), na.rm = TRUE)) %>%
  #this rounding is necessary, because the thresholds generated by DT::styleInterval() are also rounded to 4 digits after the .
  #mutate(across(starts_with("len"), round, 4)) %>%
  #mutate(mean = round(mean, 4))
  mutate(across(where(is.numeric), \(x) round(x, 4)))

n_value_cols <- ncol(df.single_nuc) - 2 # this should always be 8 as we filter longer homopolymers above

dt <- df.single_nuc %>%
  mutate(assembly = paste0('<a href="', out_dir,'/pomoxis/', assembly, '/assess_homopolymers/', assembly, '_', reference, '_count/hp_correct_vs_len.png">', assembly, '</a>')) %>%
  DT::datatable(escape = FALSE, rownames = FALSE, class="compact hover",
               options = list(pageLength = 10, searchHighlight = TRUE, paginate = TRUE, lengthChange = FALSE, fixedHeader = TRUE, order = list(list(8, 'desc'), list(0, 'asc')))) %>%
  formatStyle(names(df.single_nuc), 'font-size' = '90%') %>%
  formatRound(3:ncol(df.single_nuc), digits = 4)

color_dt(dt, 3:ncol(df.single_nuc), rep(green, n_value_cols), df.single_nuc)
```


```{r dnadiff, eval=do_reference_genomes}
asis_output("## dnadiff\n")

asis_output("dnadiff from the [mummer](https://mummer4.github.io/index.html) package compares each assembly to the reference genomes and calculates an average identity and the number of indels.\n")

filename_dnadiff_stats_tsv <- paste0(out_dir, "/dnadiff/all_stats.tsv")
df.dnadiff <- read_tsv(filename_dnadiff_stats_tsv, col_names = c("assembly", "reference", "measure", "value", "value2"), col_types = "fffdd") %>%
  pivot_wider(id_cols = c(assembly, reference), names_from = measure, values_from = value)

dt <- df.dnadiff %>%
     DT::datatable(escape = FALSE, rownames = FALSE, class="compact hover",
               options = list(pageLength = 10, searchHighlight = TRUE, paginate = TRUE, lengthChange = FALSE, fixedHeader = TRUE, order = list(list(2, 'desc'), list(3, 'asc')))) %>%
  formatStyle(names(df.dnadiff), 'font-size' = '90%') %>%
  formatRound(3, digits = 3)

color_dt(dt, c(3,4), c(green,red), df.dnadiff)
```

```{r nucdiff, eval=do_reference_genomes}
asis_output("## NucDiff\n")

asis_output("[NucDiff](https://github.com/uio-cels/NucDiff/) compares each assembly to the reference genomes and calculates the number of insertions, deletions, and substitutions.\n")

filename_nucdiff_stats_tsv <- paste0(out_dir, "/nucdiff/all_stats.tsv")

df.nucdiff <-df <- read_tsv(filename_nucdiff_stats_tsv, col_names = c("assembly", "reference", "measure", "value"), col_types = "fffd") %>%
  pivot_wider(id_cols = c(assembly, reference), names_from = measure, values_from = value)

dt <- df.nucdiff %>%
     DT::datatable(escape = FALSE, rownames = FALSE, class="compact hover",
               options = list(pageLength = 10, searchHighlight = TRUE, paginate = TRUE, lengthChange = FALSE, fixedHeader = TRUE, order = list(list(0, 'asc')))) %>%
  formatStyle(names(df.nucdiff), 'font-size' = '90%')

color_dt(dt, c(3,4,5), c(red,red, red), df.nucdiff)
```


## IDEEL

### ORF prediction with Prodigal

ORFs are predicted in each assembly using Prodigal.

```{r}
fname_prodigal_stats <- paste0(out_dir, "/ideel/prodigal_stats.tsv")

df.prodigal_stats <- read_tsv(fname_prodigal_stats, col_names = TRUE, col_types = "fcciiidi") %>%
    mutate(assembly = str_remove(file, "^.*/ideel/prodigal/")) %>%
    mutate(assembly = str_remove(assembly, "\\.faa")) %>%
    transmute(assembly, n_predicted_ORFs = num_seqs, min_len, avg_len, max_len, sum_len)

dt <- df.prodigal_stats %>%
  DT::datatable(escape = FALSE, rownames = FALSE, class="compact hover",
                options = list(pageLength = 10, searchHighlight = TRUE, paginate = TRUE, lengthChange = FALSE, fixedHeader = TRUE, order = list(list(3, 'desc'), list(0, 'asc')))) %>%
  formatStyle(names(df.prodigal_stats), 'font-size' = '90%')

color_dt(dt, 2:6, c(green, green, green, green, green), df.prodigal_stats)
```

### Comparison with Uniprot Sprot database

Predicted ORFs are searched against die Uniprot Sprot database using diamond.
Measured is the number of ORFs with a database match (n_unique_qsid), the fraction of ORFs that have the same length or ≥95% the length of the best-matching database sequence (higher fractions being better).

```{r}
dir_ideel_diamond <- paste0(out_dir, "/ideel/diamond/")
filelist <- list.files(path = dir_ideel_diamond, pattern = ".*\\.tsv", recursive = FALSE, full.names = FALSE)

df_list <- vector("list", length(filelist))
for (i in seq_along(filelist)) {
  filename <- paste0(dir_ideel_diamond, filelist[[i]])
  df <- read_tsv(filename, col_names = c("qseqid", "sseqid", "qlen", "slen"), col_types = "ccii") %>%
    mutate(assembly = filelist[[i]]) %>%
    mutate(assembly = str_remove(assembly, "\\.tsv")) %>%
    mutate(SCov = qlen / slen)

  df_list[[i]] <- df
}
df.all <- dplyr::bind_rows(df_list) %>% mutate(assembly = factor(assembly))

df.ideel_uniprot <- df.all %>%
  group_by(assembly) %>%
  summarise(
    n_unique_qsid = n_distinct(qseqid),
    n_full_length_ORFs = sum(qlen == slen),
    n_95perc_ORFs = sum(qlen >= 0.95 * slen)
  ) %>%
  ungroup()

dt <- df.ideel_uniprot %>%
  DT::datatable(escape = FALSE, rownames = FALSE, class="compact hover",
                options = list(pageLength = 10, searchHighlight = TRUE, paginate = TRUE, lengthChange = FALSE, fixedHeader = TRUE, order = list(list(2, 'desc'), list(3, 'desc'), list(0, 'asc')))) %>%
  formatStyle(names(df.ideel_uniprot), 'font-size' = '90%')

color_dt(dt, 2:4, c(green, green, green), df.ideel_uniprot)
```

```{r ideel_uniprot, warning=FALSE, fig.width = 12, fig.height = 10, out.width = '1200px'}
p <- ggplot(df.all, aes(x = SCov)) +
  geom_histogram(aes(color=assembly, fill=assembly), alpha = 0.8, position = "identity", binwidth = 0.01, size=0.2) +
  #geom_density(aes(fill = assembly), alpha = 0.4, color = "grey40") +
  facet_wrap(~assembly, scales = "free_x") +
  theme_bw() +
  ggtitle(paste("IDEEL Uniprot")) +
  theme(legend.position = "none") +
  ylab("") + xlab("qlen / slen") +
  theme(strip.text = element_text(size = 6)) +
  xlim(0, 1.5)
p
```

```{r eval=do_reference_proteomes}
asis_output("### Comparison with reference proteomes\n")

asis_output("Predicted ORFs are searched against the provided reference protein sequences using diamond.  \n")
asis_output("Measured is the number of ORFs with a database match, the fraction of ORFs that have the same length or ≥95% the length of the best-matching database sequence (higher fractions being better)")

df.ideel_ref <- data.frame(assembly = character(), reference = character(), n_seq_ref = integer(), n_predicted_ORFs = integer(), n_unique_qsid = integer(), n_unique_ssid = integer(), n_full_length_ORFs = integer(), n_min95perc_ORFs = integer())

dir_ideel_diamond_ref <- paste0(out_dir, "/ideel/diamond-ref/")
ref_list <- list.dirs(path = dir_ideel_diamond_ref, recursive = FALSE, full.names = FALSE)
for (r in seq_along(ref_list)) {
  ref <- ref_list[[r]]
  # load number of reference proteins from faa files
  len_faa_ref <- length(readAAStringSet(paste0("references-protein/", ref, ".faa")))

  filelist <- list.files(path = paste0(dir_ideel_diamond_ref,"/", ref), pattern = ".*\\.tsv", recursive = FALSE, full.names = FALSE)
  for (i in seq_along(filelist)) {
    filename <- paste0(dir_ideel_diamond_ref, "/", ref, "/", filelist[[i]])
    df <- read_tsv(filename, col_names = c("qseqid", "sseqid", "qlen", "slen"), col_types = "ccii")

    n_unique_qsid <- df %>% pull(qseqid) %>% unique() %>% length()
    n_same_len <- df %>% filter(slen == qlen) %>% nrow()
    n_min95perc_len <- df %>% filter(qlen >= 0.95 * slen) %>% nrow()
    n_unique_ssid <- df %>% pull(sseqid) %>% unique() %>% length()

    curr_assembly <- filelist[[i]] %>% str_remove("\\.tsv") %>% str_remove(paste0("_", ref))

    # load number of ORFs as predicted by prodigal in this assembly
    n_predicted_ORFs <- df.prodigal_stats %>% filter(assembly == curr_assembly) %>% pull(n_predicted_ORFs)

    df.ideel_ref <- df.ideel_ref %>% add_row(reference = ref, assembly = curr_assembly, n_seq_ref = len_faa_ref, n_predicted_ORFs = n_predicted_ORFs, n_unique_qsid = n_unique_qsid, n_unique_ssid = n_unique_ssid, n_full_length_ORFs = n_same_len, n_min95perc_ORFs = n_min95perc_len)
  }
}

dt <- df.ideel_ref %>%
  DT::datatable(escape = FALSE, rownames = FALSE, class="compact hover",
                options = list(pageLength = 10, searchHighlight = TRUE, paginate = TRUE, lengthChange = FALSE, fixedHeader = TRUE, order = list(list(0, 'asc')))) %>%
  formatStyle(names(df.ideel_ref), 'font-size' = '90%')
color_dt(dt, 6:8, c(green, green, green), df.ideel_ref)
```

```{r eval=do_bakta}
asis_output("## Bakta genome annotation\n")

asis_output("Assemblies are annotated using bakta, which searches for protein-coding and non-coding genes, as well as other genomic features.\n")

filelist <- list.files(path = path_bakta, pattern = ".*\\.txt", recursive = TRUE)

df_list <- vector("list", length(filelist))
for (i in seq_along(filelist)) {
  filename <- paste0(path_bakta, filelist[[i]])
  df <- read_delim(filename, col_names = c("label", "value"), col_types = "cc", delim = ":", skip = 9, n_max = 15, trim_ws = TRUE) %>%
    mutate(assembly = str_match(filelist[[i]],"/(.*?)$")[,2]) %>%
    mutate(assembly = str_remove(assembly, "\\.txt"))

  df_list[[i]] <- df
}
df.all <- dplyr::bind_rows(df_list) %>% mutate(assembly = factor(assembly))

df.bakta <- df.all %>%
	mutate(value = as.integer(value)) %>%
	pivot_wider(id_cols = "assembly", names_from = "label", values_from = "value")

df.bakta %>%
  DT::datatable(escape = FALSE, rownames = FALSE, class="compact hover",
                options = list(scrollX = TRUE, fixedColumns = list(leftColumns = 1), pageLength = 10, searchHighlight = TRUE, paginate = TRUE, lengthChange = TRUE, fixedHeader = TRUE, order = list(list(0, 'asc')))) %>%
  formatStyle(names(df.bakta), 'font-size' = '90%')
```


## Summary

```{r combined}
list_headings <- c("assembly")
list_base_colors <- c("0,0,0")

# merge busco and uniprot ideel
df.out <- select(df.busco, assembly, `% complete`) %>%
  left_join(select(df.ideel_uniprot, assembly, n_full_length_ORFs, n_95perc_ORFs), by = "assembly")

list_headings <- c(list_headings, "% compl.", "full ORFs", ">95% ORFs")
list_base_colors <- c(list_base_colors, green, green, green)

# join reference-protein ideel results for each reference

if(do_reference_proteomes) {
	for(i in list_ref_protein) {
		df.out <- df.out %>%
			left_join(df.ideel_ref %>% filter(reference == i) %>% select(assembly, n_full_length_ORFs, n_min95perc_ORFs), by = "assembly")

		list_headings <- c(list_headings, "full ORFs", ">95% ORFs")
		list_base_colors <- c(list_base_colors, green, green)
	}
}

# iterate through reference genomes and add columns from all tools
if(do_reference_genomes) {
	for(i in list_ref) {
		# join assess_assembly results
		df.out <- df.out %>%
			left_join(df.assess_assembly %>% filter(reference == i) %>% select(assembly, Qscore, percErr), by = "assembly")
		list_headings <- c(list_headings, "Q-score", "% Err")
		list_base_colors <- c(list_base_colors, green, red)

		# join assess_homopolymers results for each reference
		df.out <- df.out %>%
			left_join(df.single_nuc %>% filter(reference == i) %>% select(assembly, mean) %>% mutate(mean = round(mean, digits = 3)), by = "assembly")
		list_headings <- c(list_headings, "Avg Correct")
		list_base_colors <- c(list_base_colors, green)

		# join dnadiff results for each reference
		df.out <- df.out %>%
			left_join(df.dnadiff %>% filter(reference == i) %>% select(assembly, AvgIdentity, TotalIndels), by = "assembly")
		list_headings <- c(list_headings, "ANI", "Indels")
		list_base_colors <- c(list_base_colors, green, red)

		# join nucdiff results for each reference
		df.out <- df.out %>%
			left_join(df.nucdiff %>% filter(reference == i) %>% select(assembly, Insertions, Deletions, Substitutions), by = "assembly")
		list_headings <- c(list_headings, "Ins", "Del", "Sub")
		list_base_colors <- c(list_base_colors, red, red, red)
	}
}

list_th_tool_names_prot <- function(x) { htmltools::withTags(th("IDEEL reference", colspan=2)) }
list_th_tool_names <- function(x) { htmltools::withTags(list(th("assess assembly", colspan=2, style = "white-space:wrap;"), th("assess homopolymers", colspan=1, style = "white-space:wrap;"), th("dnadiff", colspan=2), th("nucdiff", colspan=3))) }

make_th_ref_prot <- function(x) { htmltools::withTags(th(x, colspan = 2)) }
make_th_ref <- function(x) { htmltools::withTags(th(x, colspan = 8)) }
make_th <- function(x) { htmltools::withTags(th(x)) }

# make a table container with custom headings
dt_container <- htmltools::withTags(table(
  class = 'display',
  thead(
    # first row has reference names
    tr(
      th(colspan = 4, "", style = "background-color: white;"),
      lapply(list_ref_protein, make_th_ref_prot),
      lapply(list_ref, make_th_ref)
    ),
    # second row has tool names
    tr(
      th("", style = "background-color: white;"),
      th("BUSCO"),
      th("IDEEL Uniprot", colspan = 2),
      lapply(list_ref_protein, list_th_tool_names_prot), # just use lapply() for calling list_th_tool_names_prot() length(list_ref_protein) times
      lapply(list_ref, list_th_tool_names)
    ),
    #third row has actual column names
    tr(
      lapply(list_headings, make_th)
    )
  )
))

dt <- df.out %>%
  DT::datatable(elementId = "dt_summary", container = dt_container, colnames = list_headings, escape = FALSE, rownames = FALSE, class="compact hover", extensions = "FixedColumns",
               options = list(scrollX = TRUE, fixedColumns = list(leftColumns = 1), pageLength = 1000, searchHighlight = TRUE, paginate = FALSE, lengthChange = FALSE, fixedHeader = TRUE, order = list(list(0, 'asc')))) %>%
  formatStyle(names(df.out), 'font-size' = '90%') %>%
  formatStyle(names(df.out), 'vertical-align' = 'top')

# to better highlight the highest value in each column:
# max value gets alpha = 1.0, lower values get alpha from 0 to 0.7
# for red colors, the alpha values should be 0 for min() and 0.2 to 1.0 for 2nd lowest to max

alpha_green <- seq(0, 0.7, length.out = 10)
alpha_red <- seq(0.3, 1.0, length.out = 10)
#i <- 11
for (i in 2:ncol(df.out)) {
	values <- df.out[[i]]
	n <- length(unique(values))
	if(n == 1) { # only one value in whole column
		list_breaks <- c(values[[1]])
		list_colors <- rep(paste0("rgba(", list_base_colors[[i]],",1.0)"), 2)
	} else {
		if(list_base_colors[[i]] == green) {
			second_highest_value <- sort(unique(values))[n-1]
			list_breaks <- seq(min(values), second_highest_value, length.out = 10)
			list_colors <- paste0("rgba(", list_base_colors[[i]],", ", c(alpha_green, 1.0), ")")
		} else {
			second_lowest_value <- sort(unique(values))[2]
			list_breaks <- seq(second_lowest_value, max(values), length.out = 10) - 0.0001
			list_colors <- paste0("rgba(", list_base_colors[[i]],", ", c(0.0, alpha_red), ")")
		}
	}

  dt <- DT::formatStyle(dt, i, backgroundColor = DT::styleInterval(list_breaks, list_colors))
}
dt
```







